### ISSUE TITLE
RequestHandler and Dependency handling

### ISSUE THREAD

**Issue by mwgustin** [Azure/azure-cosmos-dotnet-v3#3003]
Due to how RequestHandlers are initialized and used, I believe it is difficult/impossible to utilize dependency injection best practices to resolve dependencies in your RequestHandler.  

The current method of initialization forces dependencies to be resolved when building the client and the RequestHandler must essentially act as a singleton. You can see the problem even in the provided [LoggingHandler.cs](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/Microsoft.Azure.Cosmos.Samples/Usage/Handlers/LoggingHandler.cs) example.  This example creates a new TelemetryClient that's attached to the handler. This goes against the [recommendation](https://docs.microsoft.com/en-us/azure/azure-monitor/app/asp-net-core#how-can-i-track-telemetry-thats-not-automatically-collected) from App Insights to utilize a dependency injected version rather than creating a new client (obviously fine for a test/example, but in a real life application we wouldn't want to be recreating TelemetryClients and configurations everywhere). You can see a similar issue if you were to try to utilize an HttpClient where again the recommendation is to setup the client via service registration and have the application manage the instances of the client via dependency injection.

In my particular case, I want to track the RU cost over a **scoped** lifetime of an API request (example: if we make 4 CosmosDB calls to serve up the final API response, what is the total RU's for all of those calls combined), but I can't see any way with how the handlers are currently set up to allow for a scoped lifetime, or really any sort of dependency injection other than perhaps a singleton.  

Am I missing something?

**Comment by j82w** [comment]
The SDK automatically adds the RU cost of all the requests and/or retries to the final response the caller gets back.  The customer handlers that are passed in are above most of the retry logic so it will only see the original request and the final response like the caller does. 

This shows how the different handlers are linked together.
https://github.com/Azure/azure-cosmos-dotnet-v3/blob/5c718094395b73e02076980cda431c16551d1b01/Microsoft.Azure.Cosmos/src/Handler/ClientPipelineBuilder.cs#L116

**Comment by ealsur** [comment]
I understand the described limitations but assuming you are defining the CosmosClient at the same level you are defining your other DI components, can't you do:

```
services.AddSingleton<IMyService>(new MyService());
services.AddSingleton((IMyservice myservice) => {
    MyRequestHandler myRH = new MyRequestHandler(myservice);
   return new CosmosClient(......, new CosmosClientOptions() { .. adding myRH to the pipeline ..});
};
```

**Comment by mwgustin** [comment]
> The SDK automatically adds the RU cost of all the requests and/or retries to the final response the caller gets back. The custom handlers that are passed in are above most of the retry logic so it will only see the original request and the final response like the caller does.
> 
> This shows how the different handlers are linked together.
> 
> https://github.com/Azure/azure-cosmos-dotnet-v3/blob/5c718094395b73e02076980cda431c16551d1b01/Microsoft.Azure.Cosmos/src/Handler/ClientPipelineBuilder.cs#L116

It adds up the RUs to the single query response... My use case is that if an API request takes 3 separate distinct queries to fully execute (example, one pulls user information, one pulls some reference data, and then one that upserts the user after making changes), I want to be able to understand the full context of the request and store all 3 of these separate cosmos queries and the total RUs across all of them utilizing the Scoped lifetime. 



> I understand the described limitations but assuming you are defining the CosmosClient at the same level you are defining your other DI components, can't you do:
> 
> ```
> services.AddSingleton<IMyService>(new MyService());
> services.AddSingleton((IMyservice myservice) => {
>     MyRequestHandler myRH = new MyRequestHandler(myservice);
>    return new CosmosClient(......, new CosmosClientOptions() { .. adding myRH to the pipeline ..});
> };
> ```

As you say, this only works if your service has the same lifetime as your CosmosClient (ie Singleton). Otherwise you end up with captive dependencies. 

I also tried a method where I injected IServiceProvider into a singleton service and called my scoped service from there, but in ASP executions this breaks as the IServiceProvider can't understand the current scope when generated this way.

**Comment by j82w** [comment]
> It adds up the RUs to the single query response... My use case is that if an ASP.Net API request takes 3 separate distinct queries to fully execute (example, one pulls user information, one pulls some reference data, and then one that upserts the user after making changes), I want to be able to understand the full context of the request and store all 3 of these separate cosmos queries and the total RUs across all of them utilizing the Scoped lifetime.

I was only thinking of point operations which would only have 1 request and 1 response. 

For query you are correct that the SDK can/will send multiple requests depending on the query and container size. Those will have to be added together to get the total cost of the query. Each request will still add up all the retries done at lower level where retries may occur.

**Comment by ealsur** [comment]
> As you say, this only works if your service has the same lifetime as your CosmosClient (ie Singleton). Otherwise you end up with captive dependencies.

Correct, current model does not work with Transient dependency resolution because the handlers are created only once and reused.
